synapse/config/oidc_config.py

```diff
@@ -190,6 +191,19 @@ def generate_config_section(self, config_dir_path, server_name, **kwargs):
         #           When rendering, the Jinja2 templates are given a 'user' variable,
         #           which is set to the claims returned by the UserInfo Endpoint and/or
         #           in the ID Token.
+        #   It is possible to configure Synapse to only allow logins if certain attributes
+        #   match particular values in the OIDC userinfo. The requirements can be listed under
+        #   `attribute_requirements` as shown below. All of the listed attributes must
+        #   match for the login to be permitted. Note that fields listed may need to be added
+        #   to the userinfo return in the OIDC provider, by extending or adding a scope.
+        #   If the attributes listed are not part of the default scopes, these scopes may
+        #   also need to be added to the `scopes` section of the OIDC config.
+        #
+        #   attribute_requirements:
+        #     - attribute: family_name
+        #       value: "Stephensson"
+        #     - attribute: groups
+        #       value: "admin"
```

Comment:
I think it may be a good idea to stick to the default OIDC `profile` scope claims in the examples, rather than having parity with the SAML and CAS examples.
https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims

I picked `family_name` because I felt that was a reasonably common use case, restricting logins to one's family when running a homeserver, for instance. `groups` is used a lot when integrating OIDC providers with things like `oauth2-proxy`, but importantly `groups` is then a list of strings, which I have Opinions on how best to handle... (see below)

Reply:
Ah! I did not realize that is where these came from. 😄 The reasoning makes sense.